examples/httpbin-demo/src/charm.py

```diff
@@ -32,43 +32,55 @@ class HttpbinDemoCharm(ops.CharmBase):
 
     def __init__(self, framework: ops.Framework):
         super().__init__(framework)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
         framework.observe(self.on['httpbin'].pebble_ready, self._on_httpbin_pebble_ready)
         framework.observe(self.on.config_changed, self._on_config_changed)
+        self.log_level = str(self.config['log-level']).lower()  # str() is for the type checker.
+        self.container = self.unit.get_container('httpbin')
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent):
+        """Report the status of the workload (runs after each event)."""
+        if self.log_level.lower() not in VALID_LOG_LEVELS:
+            event.add_status(ops.BlockedStatus(f"invalid log level: '{self.log_level}'"))
+        try:
+            if not self.container.get_service('httpbin').is_running():
+                event.add_status(ops.MaintenanceStatus('waiting for workload'))
+        except (ops.pebble.ConnectionError, ops.ModelError):
+            event.add_status(ops.MaintenanceStatus('waiting for workload container'))
+        event.add_status(ops.ActiveStatus())
 
     def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent):
         """Define and start a workload using the Pebble API."""
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer('httpbin', self._pebble_layer, combine=True)
+        # Add initial Pebble config layer using the Pebble API.
+        self.container.add_layer('httpbin', self._pebble_layer, combine=True)
         # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        self.unit.status = ops.ActiveStatus()
+        self.container.replan()
+        # In rare cases, these calls could fail because the workload container became unavailable.
+        # If this happens, we'll let the unit go into error status. The hook will be retried, or
+        # the Juju user can investigate.
 
     def _on_config_changed(self, event: ops.ConfigChangedEvent):
         """Handle changed configuration."""
-        # Fetch the new config value
-        log_level = cast('str', self.model.config['log-level']).lower()
-
-        # Do some validation of the configuration option
-        if log_level in VALID_LOG_LEVELS:
-            # The config is good, so update the configuration of the workload
-            container = self.unit.get_container('httpbin')
-            # Push an updated layer with the new config
+        if self.log_level.lower() not in VALID_LOG_LEVELS:
+            return
+        # Update the configuration of the workload.
+        # We might not be able to access the workload container yet, so we'll try a few times.
+        max_attempts = 3
+        for attempt in range(max_attempts):
```

Comment:
Very nitpicky and possibly a matter of taste, but I'd inline `max_attempts`
```suggestion
        for attempt in range(3):
```
Oh unless you want to add it to the logging like 'attempt 1 of 3' kind of thing.

Reply:
I'm using `max_attempts` in the logging message if all attempts fail:

```python
logger.warning('Unable to reconfigure gunicorn after %d attempts.', max_attempts)
```

So I'll leave it as-is, unless you feel that this isn't necessary

Reply:
Sounds good, I didn't notice that